Few i2p fixes

mlt opened the pull request:
This works quite nice,but...
- I'm still getting dangling outstanding *on_i2p_resolve* from *http_connection* :( A call to a handler is eaten up somewhere?
- Also for some reason I'm not getting download speeds of individual peers in client_test, up speeds are fine.
- Something needs to be done with session ID generator :-)
- i2p connection has to be reestablished somehow if a router was restarted
- few asserts needs to be resolved
- <del>0.0.0.0 are saved in resume and are blocked upon restart for i2p peers.</del>

Aside of that, it works well for me with asserts disabled.

arvidn commented on src/http_connection.cpp:
```diff
@@ -373,8 +373,11 @@ void http_connection::start(std::string const& hostname, int port
 #if defined TORRENT_ASIO_DEBUGGING
 			add_outstanding_async("http_connection::on_i2p_resolve");
```
this add_outstanding_async() belongs together with the call to async_name_lookup

mlt replied:
In that case I'd have to split *on_i2p_resolve* with the case when I explicitly add a peer. What I have indeed breaks semantics, but a matching completion is called in *on_i2p_resolve*. Shall I split that function then?

arvidn commented on src/http_tracker_connection.cpp:
```diff
@@ -168,10 +168,10 @@ namespace libtorrent
-					, tracker_req().i2pconn->local_endpoint().size());
-				url += ".i2p";
+				if (tracker_req().i2pconn->local_endpoint().empty())
+					fail_impl(error_code(errors::no_i2p_endpoint), -1, "Waiting for i2p acceptor from SAM bridge", 5);
```
you're not supposed to call fail_impl() directly. call fail() instead.

mlt replied:
Is it a big no-no? I though it is the same thread etc.

arvidn commented on src/http_tracker_connection.cpp:
```diff
@@ -168,10 +168,10 @@ namespace libtorrent
+				if (tracker_req().i2pconn->local_endpoint().empty())
+					fail_impl(error_code(errors::no_i2p_endpoint), -1, "Waiting for i2p acceptor from SAM bridge", 5);
+				else
+					url += "&ip=" + tracker_req ().i2pconn->local_endpoint () + ".i2p";
```
is this required? I would have imagined it to be deduced from the connection itself

mlt replied:
No, it won't be deduced. Local *destination* is not related to anything outgoing. Another acceptor would have another base64 *destination*. There is no way to tell destination other than explicitly communicate it.

arvidn commented on src/http_tracker_connection.cpp:
```diff
@@ -333,7 +333,11 @@ namespace libtorrent
-			, tracker_req().info_hash);
+			, tracker_req().info_hash
+#if TORRENT_USE_I2P
+			, tracker_req().i2pconn != NULL
```
it would probably be better to replace the two bools (scrape_request and is_i2p) with a flags field and have two flags. That way the signature would not have to depend on TORRENT_ISE_I2P either

mlt replied:
Do you have a suggestion for a namespace to put those flags into? *request* doesn't appeal to me as in `request::scrape | request::i2p`.

arvidn commented on src/http_tracker_connection.cpp:
```diff
@@ -499,16 +507,33 @@ namespace libtorrent
-			resp.peers4.reserve(len / 6);
-			for (int i = 0; i < len; i += 6)
+#if TORRENT_USE_I2P
+			if (is_i2p) {
```
would you mind adding a test for this path too? see test/test_tracker.cpp

arvidn commented on src/escape_string.cpp:
```diff
@@ -321,15 +321,24 @@ namespace libtorrent
 			'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X',
 			'Y', 'Z', '2', '3', '4', '5', '6', '7'
 		};
+		static const char base32_table_lowercase[] =
```
would you mind adding a test exercising this code path as well? see test/test_string.cpp

arvidn commented on src/torrent.cpp:
```diff
@@ -3394,20 +3397,25 @@ namespace libtorrent
-					state_updated();
-				peers_erased(st.erased);
+				if (boost::algorithm::ends_with (i->hostname, ".b32.i2p"))
+					r.i2pconn->async_name_lookup (i->hostname.c_str ()
```
now that there's a new async_*() call, there needs to be a corresponding "add_outstanding_async" call too. The name that's passed in must match the name at the top of the handler function (in this case torrent::on_i2p_resolve()).

arvidn commented:
did you see the build failures on travis?

../src/escape_string.cpp: In function ‘std::string libtorrent::base32encode(const string&, int)’:
../src/escape_string.cpp:341:16: error: ISO C++ forbids declaration of ‘base32_table’ with no type [-fpermissive]
../src/escape_string.cpp:341:81: error: invalid initialization of reference of type ‘const int&’ from expression of type ‘const char*’
../src/escape_string.cpp:374:34: error: invalid types ‘const int[unsigned char]’ for array subscript

You also seem to have made the base32 unit test fail.

mlt commented:
Yes, I saw that. I'm changing it to a simple pointer.
> base32 unit test fail

I messed up condition for padding ;-)

arvidn commented:
if you're building with boost-build, you can pass in "--abbreviate-paths" to make the directory names shorter, and not trip windows.

arvidn commented:
I missed that you fixed the test failure. would you mind rebasing this patch on top of master so I can merge it?

mlt commented:
I rebased it. So policy->peer_list in *master* was the only thing affecting my changes?
I'd like to emphasized that with the introduction of *i2p* flag alongside with scrape, one shall be careful not to compare with *announce* flag but to check if it is not *scrape*. It may break things.
I've added another commit to avoid *yourip* in handshake. This actually made me pondering how *pex* is working? I mean how can we know *destination* for incoming connections?

arvidn commented:
I thought you received the remote end's destination when accepting an incoming i2p connection. But pex doesn't have an extension supporting non IP endpoints anyway. There would probably have to be new lists for added and dropped peers.

arvidn commented on src/http_tracker_connection.cpp:
```diff
@@ -83,7 +83,7 @@ namespace libtorrent
 	{
 		std::string url = tracker_req().url;
 
-		if (tracker_req().kind == tracker_request::scrape_request)
```
wouldn't a more reliable way (i.e. less likely to introduce errors) to make the new i2p flag not be a flag, but just a part of the enumeration, and call it i2p_announce or something like that?

arvidn commented on src/http_tracker_connection.cpp:
```diff
@@ -168,10 +168,10 @@ namespace libtorrent
-					, tracker_req().i2pconn->local_endpoint().size());
-				url += ".i2p";
+				if (tracker_req().i2pconn->local_endpoint().empty())
+					fail(error_code(errors::no_i2p_endpoint), -1, "Waiting for i2p acceptor from SAM bridge", 5);
```
shouldn't you return after calling fail()?

mlt replied:
Indeed... when I tried to return (before failing), I had some issues with the timer as it didn't start and just nothing happened. I'll add this.

arvidn commented on test/test_tracker.cpp:
```diff
@@ -101,6 +101,66 @@ TORRENT_TEST(parse_peers4)
+	if (resp.peers.size() == 11)
+	{
+		TEST_EQUAL(resp.peers[0].hostname, "wgcobfq73pzmtmcttiy2knon5bm2a7gn6j6idaiccf53ikwrecdq.b32.i2p");
+//		TEST_EQUAL(resp.peers[9].hostname, "cbmvhhhgdmxha4toqj5qhphse2n3henk6g5geev3orfxardudgza.b32.i2p");
```
is this failing? why is it here? I imagine it should probably either be made to work or be removed (or have a comment describing what needs to be done to make it pass)

mlt replied:
No, it doesn't fail. It was the shortest response I saw and I grabbed it. I feel like there is no point to test all 11 but the first and the last. I'll remove it then.

arvidn replied:
there is one more commit to this PR then, right? I think this patch is good apart from the two comments (the commented out test and returning after calling fail())

arvidn commented:
looks good

mlt commented:
> you received the remote end's destination when accepting an incoming i2p connection

This made me realize on IRC that I confused few things. You were right that a tracker should deduce our destination but for historical reasons `ip=` is required. I'll correct failure message as it is not quite correct.

Also, it seems like we do need *i2p_conn* in torrent.cpp . Having it in request is not enough. `torrent::on_torrent_download` shall use i2p connection for magnet links.

mlt commented on src/http_tracker_connection.cpp:
```diff
@@ -117,7 +117,7 @@ namespace libtorrent
 		url += escape_string((const char*)&tracker_req().info_hash[0], 20);
 
-		if (tracker_req().kind == tracker_request::announce_request)
+		if (0 == (tracker_req().kind & tracker_request::scrape_request))
```
Somehow your message didn't make it here... probably as a result of forced push...
> wouldn't a more reliable way (i.e. less likely to introduce errors) to make the new i2p flag not be a flag, but just a part of the enumeration, and call it i2p_announce or something like that?

And add i2p_scrape as well and compare for this or that? I feel like having 2 independent flags announce/scrape and is_i2p would be the least error prone. Are there many places that use announce? Maybe renaming *announce_request* to *announce_request_dontuse* is an ok solution?

arvidn replied:
I actually updated that comment a few moments after I made it basically taking back what I said :)
once I realized you wanted to combine i2p with announce. I think it makes sense with flags

mlt replied:
> you wanted to combine i2p with announce

It was your idea in the first place :sunglasses: I still suggest to alter *announce_request* flag name to make it harder to use.

str4d commented:
Hi! I2P dev here (not the one that maintains I2PSnark).

> I thought you received the remote end's destination when accepting an incoming i2p connection.

If you pass `SILENT=true` then [SAM will not send the remote end's Destination](https://geti2p.net/en/docs/api/samv3). But by default `SILENT=false`, so the first line is the B64 of the Destination.

> But pex doesn't have an extension supporting non IP endpoints anyway. There would probably have to be new lists for added and dropped peers.

I2P has a PEX implementation based on `ut_pex`, using the extension message identified as `i2p_pex` in the extension handshake. See [our BitTorrent protocol specs](https://geti2p.net/en/docs/applications/bittorrent) for details.

mlt commented:
@str4d The major showstoppers for i2p PEX are the absence of "bundled" sha-256 hash and base 64 decoding with non-standard url-friendly alphabet. The latter can be derived from boost archive iterators... the bad thing is that lookup table is not plugable there. And hopefully [hash engine from GSoC 2013](https://github.com/BoostGSoC13/boost.afio/blob/content_hashing_merge/boost/afio/hash_engine.hpp) will make it to Boost.
Also both PEX implementations shall be supported as does Vuze. It is not quite clear to me how to approach it better as the logic is same but nitty-gritty details vary. I mean whether it is possible to refactor plugin and define network specific behavior in template specialization / class descendants or something. Just ripping off an original plugin with replicated logic doesn't seem DRY. I tend to think that a single plugin without much refactoring should handle both networks and include appropriate messages according to extended handshake, i.e. i2p and/or clearnet.
Taking all that into considerations, it is highly likely it won't be a part of this particular pull request. It is just a bare minimum for now.

arvidn commented:
I would be happy to merge this pull request with the last two comments fixed (which I think are minor). Also, please rebase it on top of master

arvidn commented:
thanks!
